### ISSUE TITLE
Allow to recover an instance with a new uuid.

### ISSUE THREAD

**Issue by sergepetrenko** [tarantool/tarantool#4743]
Having an anonymous replica it is now possible to add a new member to the cluster by just copying one of the masters' `.xlog` and `.snap` files to a fresh directory, replace the instance uuids stored in both files with some new value, recover the instance in this fresh directory as an anonymous replica,
```lua
box.cfg{replication_anon=true, read_only=true, replication={...}}
```
and then promote it to a normal one with
```lua
box.cfg{replication_anon=false}
```
This seems to be faster than starting an empty instance and replicating all the data from an existing master, and also allows to get an exact copy of a remote instance, so maybe it's worth allowing to recover an instance from someone else's `.xlog` and `.snap` files, when `instance_uuid` is set. This will make the process easier.

**Comment by Totktonada** [comment]
We also need to document this way to bootstrap a replica.

**Comment by kyukhin** [comment]
> We also need to document this way to bootstrap a replica.

File a doc ticket pls.

**Comment by Totktonada** [comment]
> > We also need to document this way to bootstrap a replica.
> 
> File a doc ticket pls.

I worth to do after the change Serge proposed if we'll decide to implement it.

**Comment by kostja** [comment]
This used to be the case in 1.5. And there were a lot of support cases from DBAs doing the procedure manually and screwing themselves up. 
The reason one can't bootstrap an instance by moving the data is for a reason: to not allow people to screw themselves up by creating malformed clusters.

**Comment by sergepetrenko** [comment]
@kostja Maybe we do not allow such behaviour by default, but add a separate `box.cfg` option similar to `force_recovery`, like `force_instance_uuid=true` or something? This'll give the person performing such a bootstrap an impression that they're doing this at their own risk

**Comment by sergepetrenko** [comment]
I believe what you're saying was the default behaviour in 1.5. Am I wrong?

**Comment by cyrillos** [comment]
After thinking a lot about this feature the conclusion I took is NAK. This ruins files integrity and rather force us to support this brain damaged logic forever. Currently tarantool assumes that uuids are stoned in log\sap files. And this must never be changed.

In turn I consider implementing third-party tool (or module) to convert snap\xlog files to update uuids as a workaround but again this won't be officially supported functionality and may be dropped off at any moment.

**Comment by cyrillos** [comment]
I've a number of discussions about this functionality. We are not going to implement it on a core level since this is simply wrong from an architectural point of view and will force us to carry backward compatibility forever (why real issue is on the vinyl level).

I prepared the commit 23637f5f88c3e63f8cd508dff459cf76793b6a22 just for the reference where we can migrate the images via `tarantoolctl` script. Once can use it on own risk but we never promise any guarantee.

During investigation of this issue I've simplified and cleaned up code a bit (commits are already merged) so time was not wasted for nothing.

**Comment by kyukhin** [comment]
Let's close it then.

**Comment by cyrillos** [comment]
Since we dropped the patch and branch the commit is vanished. I just pin it here in comments in case someone needs it in future.
```
From 23637f5f88c3e63f8cd508dff459cf76793b6a22 Mon Sep 17 00:00:00 2001
From: Cyrill Gorcunov <gorcunov@gmail.com>
Date: Thu, 25 Jun 2020 11:24:40 +0300
Subject: [PATCH] extra/dist/tarantoolctl: implement ability to change UUIDs in WAL

We've a number of complains that adding a new replica into vinyl-based
cluster consume too much time especially when database is quite big in
size (number of gigabytes).

To workaround this people just copy WAL files (ie .xlog/.snap/.vylog, ...)
from the master node to a new place, change the instance UUID then restart
the replica. And this process is faster than "proper" restore from empty
replica.

We know that correct fix is to figure out where exactly vynil getting
slow down and improve this code but it still requires too much
investigation time.

Lets provide some automation for known workaround instead: one can
run "tarantoolctl meta --to_uuid=some-new-uuid PATH" and UUIDs get
converted to a new value, then replica can be restarted and connected
to the cluster.

This is ugly and dirty hack, such funtionality will never be supported
inside tarantool core core itself and won't be ever recommended for use.

Moreover we definitely will revert this commit in (hopefully near)
future, thus the commit is just to automate actions.

Example: when run without keywords arguments it will print UUIDs found

 | tarantoolctl meta ../dumper/examples/replication/master
 | meta: parsing ../dumper/examples/replication/master/00000000000000000000.xlog
 | meta: read 1024 bytes
 | meta: UUID found 0a737b86-cffb-4b02-b028-05daa9906fd3
 | meta: parsing ../dumper/examples/replication/master/00000000000000000017.xlog
 | meta: read 123 bytes
 | meta: UUID found 0a737b86-cffb-4b02-b028-05daa9906fd3
 | meta: parsing ../dumper/examples/replication/master/00000000000000000000.snap
 | meta: read 1024 bytes
 | meta: UUID found 0a737b86-cffb-4b02-b028-05daa9906fd3
 | meta: parsing ../dumper/examples/replication/master/00000000000000000000.vylog
 | meta: read 236 bytes
 | meta: UUID found 0a737b86-cffb-4b02-b028-05daa9906fd3

Lets change UUID 0a737b86-cffb-4b02-b028-05daa9906fd3 to some new value
b741980f-3782-418c-8d5f-d6c3d75e7eca

 | tarantoolctl meta ../dumper/examples/replication/master --from-uuid="0a737b86-cffb-4b02-b028-05daa9906fd3" --to-uuid="b741980f-3782-418c-8d5f-d6c3d75e7eca"
 | meta: parsing ../dumper/examples/replication/master/00000000000000000000.xlog
 | meta: read 1024 bytes
 | meta: UUID found 0a737b86-cffb-4b02-b028-05daa9906fd3
 | meta: Changing UUIDs 0a737b86-cffb-4b02-b028-05daa9906fd3 -> b741980f-3782-418c-8d5f-d6c3d75e7eca
 | meta: parsing ../dumper/examples/replication/master/00000000000000000017.xlog
 | meta: read 123 bytes
 | meta: UUID found 0a737b86-cffb-4b02-b028-05daa9906fd3
 | meta: Changing UUIDs 0a737b86-cffb-4b02-b028-05daa9906fd3 -> b741980f-3782-418c-8d5f-d6c3d75e7eca
 | meta: parsing ../dumper/examples/replication/master/00000000000000000000.snap
 | meta: read 1024 bytes
 | meta: UUID found 0a737b86-cffb-4b02-b028-05daa9906fd3
 | meta: Changing UUIDs 0a737b86-cffb-4b02-b028-05daa9906fd3 -> b741980f-3782-418c-8d5f-d6c3d75e7eca
 | meta: parsing ../dumper/examples/replication/master/00000000000000000000.vylog
 | meta: read 236 bytes
 | meta: UUID found 0a737b86-cffb-4b02-b028-05daa9906fd3
 | meta: Changing UUIDs 0a737b86-cffb-4b02-b028-05daa9906fd3 -> b741980f-3782-418c-8d5f-d6c3d75e7eca

Now the images are representing instance with new uuid and ready to be reconnected to
a cluster as an anonymous replica first, then propagated to a regular replica.

Fixes #4743

Signed-off-by: Cyrill Gorcunov <gorcunov@gmail.com>
---
 extra/dist/tarantoolctl.in | 132 +++++++++++++++++++++++++++++++++++++
 1 file changed, 132 insertions(+)

diff --git a/extra/dist/tarantoolctl.in b/extra/dist/tarantoolctl.in
index a98c61b52..ecc25f269 100755
--- a/extra/dist/tarantoolctl.in
+++ b/extra/dist/tarantoolctl.in
@@ -889,6 +889,110 @@ local function cat()
     end
 end
 
+local function meta()
+    local opts = keyword_arguments
+    local dir = table.remove(positional_arguments, 1)
+
+    if dir == nil then
+        error("Empty DIR is provided")
+    end
+
+    local function parse_meta(file_name)
+        log.info("meta: parsing %s", file_name)
+
+        -- Make sure the file is big enough to
+        -- carry instance id (at least EOF mark
+        -- should be present which is 4 bytes).
+        local st, err = fio.stat(file_name)
+        if err ~= nil then
+            log.error("meta: Unable to fio.stat('%s'): %s", file_name, err)
+            error("Can't fetch file stat")
+        end
+        if st.size < 4 then
+            log.error("meta: file size is less than EOF marker")
+            error("File size is too small")
+        end
+
+        -- 1K buffer should be enough to carry the whole
+        -- meta information block.
+        local meta_size = st.size
+        if meta_size > 1024 then
+            meta_size = 1024
+        end
+
+        local fh, err = fio.open(file_name, {'O_RDWR'})
+        if err ~= nil then
+            log.error("meta: Unable to fio.open('%s'): %s", file_name, err)
+            error("Can't open file")
+        end
+
+        local data = fh:read(meta_size)
+        log.info("meta: read %d bytes", #data)
+
+        local i, j = string.find(data, "Instance: ")
+        if i == nil then
+            log.error("meta: No instance record found in %s", file_name)
+            fh:close()
+            error("No instance record found")
+        end
+
+        local ii, jj = string.find(data, "\n", i + 1)
+        if ii == nil then
+            log.error("meta: No newline after instance found in %s",
+                      file_name)
+            fh:close()
+            error("No newline after instance found")
+        end
+
+        local uuid_old = string.sub(data, j + 1, jj - 1)
+        if #uuid_old ~= 36 then
+            log.info("meta: UUID length mismatch (%d bytes)", #uuid_old)
+            fh:close()
+            error("UUID length mismatch")
+        end
+
+        -- If no keyword_arguments present we just print
+        -- out UUIDs detected for info sake.
+        log.info("meta: UUID found %s", uuid_old)
+
+        if keyword_arguments['from-uuid'] ~= nil then
+            if uuid_old ~= keyword_arguments['from-uuid'] then
+                log.info("meta: UUIDs mismatch (%s ~= ~%s)",
+                         keyword_arguments['from-uuid'], uuid_old)
+                fh:close()
+                error("UUIDs mismatch")
+                end
+            end
+
+        if keyword_arguments['to-uuid'] ~= nil then
+            local uuid_new = keyword_arguments['to-uuid']
+            log.info("meta: Changing UUIDs %s -> %s",
+                     uuid_old, uuid_new)
+            if not fh:pwrite(uuid_new .. "\n", i + 9) then
+                log.error("meta: Unable to write new UUID %s", uuid_new)
+                error("Unable to write new UUID")
+            end
+        end
+
+        fh:close()
+    end
+
+    local xlogs = fio.glob(dir .. '/*.xlog')
+    for i=1,#xlogs do
+        parse_meta(xlogs[i])
+    end
+
+    local snaps = fio.glob(dir .. '/*.snap')
+    for i=1,#snaps do
+        parse_meta(snaps[i])
+    end
+
+    local vylogs = fio.glob(dir .. '/*.vylog')
+    for i=1,#vylogs do
+        parse_meta(vylogs[i])
+    end
+end
+
 local function play()
     local opts = keyword_arguments
     local uri = table.remove(positional_arguments, 1)
@@ -1160,6 +1264,30 @@ local commands = setmetatable({
             weight = 90,
             deprecated = false,
         }
+    }, meta = {
+        func = exit_wrapper(meta), process = process_remote, help = {
+            header =
+                "%s meta DIR.. [--from-uuid=from_uuid] [--to-uuid=to_uuid]",
+            description =
+[=[
+        Change instance UUID in .snap/.xlog/.vylog files to another UUID.
+
+        The only purpose of this operation is to speedup procedure
+        of adding a new replica into a cluster. Directory with files
+        must be copied from the master node only.
+
+        This functionality is temporary solution and rather an ugly hack.
+        It is not guaranteed to be kept forever and may be dropped off
+        in any moment in future. And there is absolutely no guarantee
+        for correctness of results.
+
+        Supported options:
+        * --from-uuid=from_uuid instance UUID to be changed from.
+        * --to-uuid=to_uuid instance UUID to be set.
+]=],
+            weight = 100,
+            deprecated = false,
+        }
     }, play = {
         func = exit_wrapper(play), process = process_remote, help = {
             header =
@@ -1271,6 +1399,8 @@ local function populate_arguments()
         ka                = ka                or {}
         ka.from           = ka.from           or 0
         ka.to             = ka.to             or -1ULL
+        ka['from-uuid']   = ka['from-uuid']   or nil
+        ka['to-uuid']     = ka['to-uuid']     or nil
         ka['show-system'] = ka['show-system'] or false
         ka.format         = ka.format         or 'yaml'
         return ka
@@ -1295,6 +1425,8 @@ local function populate_arguments()
         { 'show-system', 'boolean' },
         { 'from',        'number'  },
         { 'to',          'number'  },
+        { 'from-uuid',   'string'  },
+        { 'to-uuid',     'string'  },
         { 'help',        'boolean' },
         { 'format',      'string'  },
         { 'replica',     'number+' },
-- 
2.26.2

```
